src/script/interpreter.cpp

```diff
@@ -1444,19 +1544,60 @@ void PrecomputedTransactionData::Init(const T& txTo, std::vector<CTxOut>&& spent
         m_spent_scripts_single_hash = GetSpentScriptsSHA256(m_spent_outputs);
         m_bip341_taproot_ready = true;
     }
+
+    if (force) {
+        // Eagerly compute BIP119 state
+        BIP119EagerInit(txTo);
+        // Disable Lazy Sync Wrapper
+        m_bip119_cache_synchronizer = nullptr;
+    } else if (!m_bip119_cache_synchronizer) {
+        // If no function was provided, assume that we are single-threaded but
+        // still cache-on-first-use
+        auto single_threaded = [&](std::function<void()> f)
+        {
+            f();
+            // disables future calls from recomputing
+            m_bip119_cache_synchronizer = nullptr;
+        };
+        m_bip119_cache_synchronizer = single_threaded;
```

Comment:
Shouldn't this be:

```c++
m_bip119_c_s = (f != nullptr ? f : single_threaded);
```

or something? Otherwise nothing seems to be done with the `bip_119_cache_synchronizer_t f` that was passed in other than sometimes asserting that it's `nullptr`.

Reply:
https://github.com/bitcoin/bitcoin/pull/21702/files/c24b553d3fe7db19c8c37934a196334911412a0e#diff-a0337ffd7259e8c7c9a7786d6dbd420c80abfa1afdb34ebae3261109d9ae3c19R1589 this line?

Reply:
No? I'm looking at the `Init()` function, your link is something that calls `Init()` with `f` defaulted to `nullptr`?

https://github.com/bitcoin/bitcoin/pull/21702/files/c24b553d3fe7db19c8c37934a196334911412a0e#diff-a0337ffd7259e8c7c9a7786d6dbd420c80abfa1afdb34ebae3261109d9ae3c19R1496

Reply:
https://github.com/bitcoin/bitcoin/pull/21702/files/c24b553d3fe7db19c8c37934a196334911412a0e#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R1736-R1737 seems to be the only place where Init's `f` parameter isn't defaulted.

Reply:
I think you're right that this is broken -- do you have a test condition that triggers this?

Very good catch, I think this should be eligible for the bipbounty.org bounty as a Major?

cc @jamesob can you re-do the benchmarks so we can also see if it'd be reasonable to remove these patches from the PR still?

Reply:
@JeremyRubin yup, I'll do that this week.

Reply:
update here:

I've been trying to figure out how/why this was not making a test failure happen, and I think I know how to trigger it with a test case (but haven't written one out). To trigger it, all of the following must be true:

Things that can be set up deterministically:
1) There is no cached evaluation of the script from a single threaded run (which can happen in the mempool), so this should be tested with a direct block submission
2) 143 and 341 must be disabled OR there must be legacy inputs that have non-empty scriptsigs (to trigger a hash getting computed).
3) There is more than one CTV Input
4) The node is running with multiple workers

Things that happen by chance (don't know exactly how to trigger this part via a test):

5) The scriptcheck jobs get scheduled on two different checkqueue workers.
6) There is a data-race on two inputs such that they don't happen to correctly synchronize (CheckQueue has some locks in it)
7) The data-races don't happen to both observe the fully written value -- since the cache is deterministic, on most architectures the cache computations won't result in a "garbage value", since the write is the same. So when the single threaded computation finishes, the values will *always* be consistent. This is aided by the fact that the last write to the synchronizer single threaded is to set nullptr for the function, so we either do parallel identical runs, or we do one run and the other observes it.


7 is really the criticial part. Even if we end up using the single threaded initialization, it seems that it does not (on most architectures?) lead to an actual race, no matter what.

This isn't to say that this shouldn't get patched -- it's super clearly not formally correct -- but also that it seems not possible to actually write a failing test for, which is what I would normally do under the circumstances.

Reply:
In any case, patch fixing the issue is here: https://github.com/bitcoin/bitcoin/pull/21702/commits/d4af88d25dd5e5d8196988bf00c2dcefcc5d96b4

Reply:
I think that means that you should be able to trigger a tsan warning by having a tx with, say, 100 inputs, 10 of which are plain p2pkh, and the remaining 90 are different ctv inputs?

Reply:
i think it should be catchable under tsan (not entirely familiar with what it can and can't catch), but it will require some tuning or a lot of runs to actually get it to trigger? I'll see if I can put something together at some point to test the pre-patch post-patch behaviors.

would you prefer leaving this unresolved till such a test exists?

Reply:
Here's a test, I think: https://github.com/ajtowns/bitcoin/commit/5beeb28bde367769e189a7ff5857a190bc19c15b

When I run it under tsan without your patch, I get: "[node 1] Node returned non-zero exit code (66) when stopping"; with errors in the logs that (after lots of trimming) look like:

```
WARNING: ThreadSanitizer: data race (pid=993218)
  Read of size 8 at 0x7bb40001ce58 by thread T7:
    #6 uint256 (anonymous namespace)::GetDefaultCheckTemplateVerifyHashEmptyScript<CTransaction>(CTransaction const&, uint256 const&, uint256 const&, unsigned int) src/./src/script/interpreter.cpp:1463:9 (bitcoind+0x11e0fd2) (BuildId: ec0
7d2273791a5c68cd1432ff17dfea367618622)

  Location is heap block of size 36720 at 0x7bb400014000 allocated by thread T17:
    #7 CChainState::ConnectBlock(CBlock const&, BlockValidationState&, CBlockIndex*, CCoinsViewCache&, bool) src/./src/validation.cpp:2180:45 (bitcoind+0x8d4595) (BuildId: ec07d2273791a5c68cd1432ff17dfea367618622)

  Location is heap block of size 36720 at 0x7bb400014000 allocated by thread T17:
    #1 std::__new_allocator<PrecomputedTransactionData>::allocate(unsigned long, void const*) /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/new_allocator.h:137:27 (bitcoind+0x95673e) (BuildId: ec07d2273791a5c68cd1432ff17dfea367618622)

  Thread T7 'b-scriptch.6' (tid=993226, running) created by main thread at:
  Thread T4 'b-scriptch.3' (tid=993223, running) created by main thread at:
  Thread T17 'b-msghand' (tid=993243, running) created by main thread at:
```

configure options are `./configure --with-gui=no CC=clang CXX=clang++ --with-incompatible-bdb --enable-zmq --enable-debug --enable-werror --with-sanitizers=thread --enable-suppress-external-warnings --enable-experimental`

With your patch, it passes and there aren't any ThreadSanitizer entries in the log.

Reply:
condition 2 i think was incorrect, because the line:
```
 m_scriptSigs_single_hash = NoScriptSigs(txTo) ? uint256{} : GetScriptSigsSHA256(txTo);
```

which could be written instead as 
```
if (!NoScriptSigs(txTo)) {
    m_scriptSigs_single_hash = GetScriptSigsSHA256(txTo);
}
```

and then the condition 2 would be accurate, I think. But because it's using `?` it always assigns, even though the new value is the same as the zero initialize.